homeassistant/components/freedompro/sensor.py

```diff
@@ -0,0 +1,100 @@
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Freedompro sensor."""
+    api_key = entry.data[CONF_API_KEY]
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        Device(hass, api_key, device, coordinator)
+        for device in coordinator.data
+        if device["type"] == "temperatureSensor"
+        or device["type"] == "humiditySensor"
+        or device["type"] == "lightSensor"
+    )
+
+
+class Device(CoordinatorEntity, SensorEntity):
+    """Representation of an Freedompro sensor."""
+
+    def __init__(self, hass, api_key, device, coordinator):
+        """Initialize the Freedompro sensor."""
+        super().__init__(coordinator)
+        self._hass = hass
+        self._session = aiohttp_client.async_get_clientsession(self._hass)
+        self._api_key = api_key
+        self._attr_name = device["name"]
+        self._attr_unique_id = device["uid"]
+        self._type = device["type"]
+        self._characteristics = device["characteristics"]
+        self._attr_device_info = {
+            "name": self._attr_name,
+            "identifiers": {
+                (DOMAIN, self._attr_unique_id),
+            },
+            "model": self._type,
+            "manufacturer": "Freedompro",
+        }
+        self._attr_device_class = (
+            DEVICE_CLASS_TEMPERATURE
+            if self._type == "temperatureSensor"
+            else DEVICE_CLASS_HUMIDITY
+            if self._type == "humiditySensor"
+            else DEVICE_CLASS_ILLUMINANCE
+        )
+        self._attr_state_class = (
+            STATE_CLASS_MEASUREMENT
+            if self._type == "temperatureSensor"
+            else STATE_CLASS_MEASUREMENT
+            if self._type == "humiditySensor"
+            else None
+        )
+        self._attr_unit_of_measurement = (
+            TEMP_CELSIUS
+            if self._type == "temperatureSensor"
+            else PERCENTAGE
+            if self._type == "humiditySensor"
+            else LIGHT_LUX
+        )
+        self._attr_state = 0
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        device = next(
+            (
+                device
+                for device in self.coordinator.data
+                if device["uid"] == self._attr_unique_id
+            ),
+            None,
+        )
+        if device is not None and "state" in device:
+            state = device["state"]
+            if "currentAmbientLightLevel" in state:
+                self._attr_state = state["currentAmbientLightLevel"]
+            if "currentRelativeHumidity" in state:
+                self._attr_state = state["currentRelativeHumidity"]
+            if "currentTemperature" in state:
+                self._attr_state = state["currentTemperature"]
```

Comment:
Based on `self._type` we should know which key to test for, instead of trying them all?